Memoize test results

Janiczek opened the pull request:
Relates to #118 and #112.

Some shrinkers - notably, list shrinker from my other pull request (Fix list shrinker) - have a huge number of possible paths to take, many of them shared (ie. every shortening of a list eventually tests the empty list).

This commit makes the `fuzzTest` function (which is responsible for shrinking in case of failure, etc.) remember the test results in between runs and reuse them when encountering the values again.

This optimization makes the fuzz tests 2x to 30x faster...

...although some care is required before merging this with regards to bugs (does it work correctly?) and it could very possibly be refactored some more. I expect to add more commits doing that to this pull request.

Janiczek commented:
To consider:
- [ ] **memory usage.** Now it memoizes every tested value, even one that passes the tests. Does that make the process memory-hungry? If we only memoized failing values, that would be a tradeoff in number of skipped results vs. memory usage over the course of the run.
- [ ] **correctness.** False negatives, false positives. Does it catch everything the previous code caught?
- [ ] **code readability.** Can it be DRY-ed some more, refactored, ...?
- [ ] **functionality.** Does it do what it says it does? (ie. when shrinking, does it correctly memoize each semi-result? Or only eg. the minimal failure, which would be a bug)

I'm currently testing the last one - functionality.

Janiczek commented:
Do you know of some diverse test suites publicly available on GitHub? It would be good to test the new code on something nontrivial. I would then make benchmarks from these test suites.

Janiczek commented:
Suggested test suites to try (I will probably prioritize bigger ones and more fuzz-based ones)
- https://github.com/rtfeldman/elm-css
- https://github.com/gampleman/form-of-address
- https://github.com/krisajenkins/elm-exts
- https://github.com/mgold/elm-geojson
- https://github.com/rohanorton/Purely-Functional-Data-Structures
- https://github.com/truqu/elm-dictset
- https://github.com/folkertdev/elm-deque
- https://github.com/newlandsvalley/elm-binary-base64

I have a question regarding the benchmarking: **what makes the benchmark useful?** Is it the average running times, std.dev, variance, ops/sec, ...? Ie. **what question are we trying to answer?**

I have a feeling that we should test the **50th percentile** = median (typical use case) and maybe a **95th percentile** (worst-case) - and comparing the latest tagged `elm-test` version with latest+this PR.

Related to that: would 20 runs for each variant be enough?

Thoughts?

Janiczek commented:
So, the benchmarks are here:

- [table](https://github.com/elm-community/elm-test/files/768070/output.txt)
- [code](https://github.com/elm-community/elm-test/files/768078/benchmarks.zip)

I only ran **fuzz tests** in the repos that I cloned and benchmarked (unit tests are commented out).

Throw it into an Excel and voila:

![selection_095](https://cloud.githubusercontent.com/assets/149425/22848253/7517cb36-eff3-11e6-877e-edfa82383922.png)

So, for me this means two things: find out what happens in **elm-css** and **elm-geojson** (with the more granular output of `elm-test --report json` showing duration for every test). The rest looks pretty good!

- [ ] elm-css
- [ ] elm-geojson

What do you think?

Janiczek commented:
By the way, I think this pull request is ready for merging (or code review).

I briefly looked at the elm-css tests and while I'm not sure, I think the overhead just outweighs the duration of test itself.

When can we merge this (and #118)?

rtfeldman commented:
I'm happy to merge both. @mgold what do you think?

mgold commented on src/Test/Internal.elm:
```diff
@@ -53,40 +58,55 @@ fuzzTest fuzzer desc getExpectation =
+                        -- we can skip this, already have the result in `failures`
+                        failures
+
+                    Nothing ->
```
I suspect we should be extracting some functions here, and possibly even record types, but that's out of scope of this PR.

mgold commented on src/Test/Internal.elm:
```diff
@@ -99,28 +119,65 @@ fuzzTest fuzzer desc getExpectation =
     in
-        Dict.insert (toString result) finalExpectation dict
+        ( Dict.insert (toString minimalValue) finalExpectation failures
+        , Dict.insert (toString failingValue) (ShrinksTo minimalValue) results
```
Is it correct to use `results` and not `results_` here? Could the underscore be changed to `newResults` or something more meaningful?

mgold commented:
I'd like an answer to my second comment (the first is more a mental note).

Janiczek commented:
@mgold I have renamed the parameters to be more descriptive; and I have fixed the usage of `newResults` - thanks for the catch!

When I've run new benchmarks, the changes are negligible.

mgold commented:
...and that is why we use descriptive variable names.
